File: backend/src/test/java/org/globe42/web/charges/ChargeControllerMvcTest.java
package org.globe42.web.charges;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.math.BigDecimal;
import java.util.Optional;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.globe42.dao.ChargeDao;
import org.globe42.dao.ChargeTypeDao;
import org.globe42.dao.PersonDao;
import org.globe42.domain.Charge;
import org.globe42.domain.ChargeType;
import org.globe42.domain.Person;
import org.globe42.test.Answers;
import org.globe42.test.GlobeMvcTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC tests for {@link ChargeController}
 * @author JB Nizet
 */
@GlobeMvcTest(ChargeController.class)
public class ChargeControllerMvcTest {
    @MockBean
    private PersonDao mockPersonDao;

    @MockBean
    private ChargeDao mockChargeDao;

    @MockBean
    private ChargeTypeDao mockChargeTypeDao;

    @Autowired
    private MockMvc mvc;

    @Autowired
    private ObjectMapper objectMapper;

    private Person person;

    @BeforeEach
    public void prepare() {
        person = new Person(42L);
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));
    }

    @Test
    public void shouldList() throws Exception {
        Charge charge = ChargeControllerTest.createCharge(12L);
        person.addCharge(charge);

        mvc.perform(get("/api/persons/{personId}/charges", person.getId()))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(charge.getId()))
           .andExpect(jsonPath("$[0].monthlyAmount").value(charge.getMonthlyAmount().doubleValue()))
           .andExpect(jsonPath("$[0].type.id").value(charge.getType().getId().intValue()));
    }

    @Test
    public void shouldDelete() throws Exception {
        Charge charge = ChargeControllerTest.createCharge(12L);
        person.addCharge(charge);

        when(mockChargeDao.findById(charge.getId())).thenReturn(Optional.of(charge));

        mvc.perform(delete("/api/persons/{personId}/charges/{chargeId}", person.getId(), charge.getId()))
           .andExpect(status().isNoContent());
    }

    @Test
    public void shouldCreate() throws Exception {
        ChargeType chargeType = ChargeControllerTest.createChargeType(12L);

        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));
        when(mockChargeDao.save(any(Charge.class)))
            .thenAnswer(Answers.<Charge>modifiedFirstArgument(charge -> charge.setId(345L)));

        ChargeCommandDTO command = new ChargeCommandDTO(chargeType.getId(), BigDecimal.TEN);
        mvc.perform(post("/api/persons/{personId}/charges", person.getId())
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.id").value(345));
    }
}

File: backend/src/test/kotlin/org/globe42/web/charges/ChargeControllerMvcTest.kt
package org.globe42.web.charges

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.globe42.dao.ChargeDao
import org.globe42.dao.ChargeTypeDao
import org.globe42.dao.PersonDao
import org.globe42.domain.Charge
import org.globe42.domain.Person
import org.globe42.test.GlobeMvcTest
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.delete
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import java.math.BigDecimal

/**
 * MVC tests for [ChargeController]
 * @author JB Nizet
 */
@GlobeMvcTest(ChargeController::class)
class ChargeControllerMvcTest(
    @Autowired private val mvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {
    @MockkBean
    private lateinit var mockPersonDao: PersonDao

    @MockkBean(relaxUnitFun = true)
    private lateinit var mockChargeDao: ChargeDao

    @MockkBean
    private lateinit var mockChargeTypeDao: ChargeTypeDao

    private lateinit var person: Person

    @BeforeEach
    fun prepare() {
        person = Person(42L)
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person
    }

    @Test
    fun `should list`() {
        val charge = createCharge(12L)
        person.addCharge(charge)

        mvc.get("/api/persons/{personId}/charges", person.id).andExpect {
            status { isOk }

            jsonValue("$[0].id", charge.id!!)
            jsonValue("$[0].monthlyAmount", charge.monthlyAmount.toDouble())
            jsonValue("$[0].type.id", charge.type.id!!.toInt())
        }
    }

    @Test
    fun `should delete`() {
        val charge = createCharge(12L)
        person.addCharge(charge)

        every { mockChargeDao.findByIdOrNull(charge.id!!) } returns charge

        mvc.delete("/api/persons/{personId}/charges/{chargeId}", person.id, charge.id).andExpect {
            status { isNoContent }
        }
    }

    @Test
    fun `should create`() {
        val chargeTypeId = 12L
        val chargeType = createChargeType(chargeTypeId)

        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns chargeType
        every { mockChargeDao.save(any<Charge>()) } answers { arg<Charge>(0).apply { id = 345L } }

        val command = ChargeCommandDTO(chargeTypeId, BigDecimal.TEN)
        mvc.post("/api/persons/{personId}/charges", person.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isCreated }
            jsonValue("$.id", 345)
        }
    }
}
